ibridges/irodsconnector/sync.py

```diff
@@ -0,0 +1,356 @@
+
+    The `max_level` option controls the depth up to which the file tree will
+    be synchronized. With `max_level` set to None (default), there is no limit
+    (full recursive synchronization). A max level of 1 synchronizes only the
+    source's root, max level 2 also includes the first set of
+    subfolders/subcollections and their contents, etc.
+
+    The `copy_empty_folders` option controls whether folders/collections that
+    contain no files or subfolders/subcollections will be synchronized (default
+    False).
+
+    The `dry_run` option lists all the source files and folders that need to
+    be synchronized without actually performing the synchronization.
+
+    The `verify_checksum` option will calculate and verify the checksum on the
+    data after up- or downloading. A checksum mismatch will generate an error,
+    but will not abort the synchronization process (default True).
+
+    Lastly, `session` requires an authorized instance of `ibridges.Session`.
+    """
+
+    assert isinstance(source, IrodsPath) or isinstance(target, IrodsPath), \
+        "Either source or target should be an iRODS path."
+    assert not (isinstance(source, IrodsPath) and isinstance(target, IrodsPath)), \
+        "iRODS to iRODS copying is not supported."
+
+    log.info("Syncing '%s' --> '%s'%s", source, target, ' (dry run)' if dry_run else '')
+
+    if isinstance(source, IrodsPath):
+        assert source.collection_exists(), \
+            f"Source collection '{source.absolute_path()}' does not exist"
+        src_files, src_folders=_get_irods_tree(
+            coll=get_collection(session=session, path=source),
+            max_level=max_level,
+            ignore_checksum=ignore_checksum)
+    else:
+        assert Path(source).is_dir(), f"Source folder '{source}' does not exist"
+        src_files, src_folders=_get_local_tree(
+            path=source,
+            max_level=max_level,
+            ignore_checksum=ignore_checksum)
+
+    if isinstance(target, IrodsPath):
+        assert target.collection_exists(), \
+            f"Target collection '{target.absolute_path()}' does not exist"
+        tgt_files, tgt_folders=_get_irods_tree(
+            coll=get_collection(session=session, path=target),
+            max_level=max_level,
+            ignore_checksum=ignore_checksum)
+    else:
+        assert Path(target).is_dir(), f"Target folder '{target}' does not exist"
+        tgt_files, tgt_folders=_get_local_tree(
+            path=target,
+            max_level=max_level,
+            ignore_checksum=ignore_checksum)
+
+    # compares the relative paths
+    folders_diff=sorted(
+        set(src_folders).difference(set(tgt_folders)),
+        key=lambda x: (x.path.count('/'), x.path))
+
+    # compares the checksum, file size, file name & relative path
+    files_diff=sorted(
+        set(src_files).difference(set(tgt_files)),
+        key=lambda x: (x.path.count('/'), x.path))
+
+    if isinstance(target, IrodsPath):
```

Comment:
Don't know if there are many use csaes, but maybe there's research groups that share a folder on YODA and work locally on individual machines. It would be quite easy to implement.